### ISSUE TITLE
Feature request: More intelligent album matching when only searching on a few tracks

### ISSUE THREAD

**Issue by RollingStar** [beetbox/beets#2454]
### Problem

Beets sometimes imports some of an album but misses a track or two. Then it starts processing the last few tracks, and panics from not having the rest of the album.

### Proposed Fix

Have the importer look in the existing library and increase a candidate's match score if the tracks would fit in with the tracks of the library. I.e., if the library has track 1-10 of an album and Beets is trying to find where tracks 11-13 of an album go, it shouldn't only give a 30% score to a choice that matches with the rest of the user's library.

Example of this problem. Beet's first guess was already in my library!

"This album is already in the library!
Old: 12 items, MP3, 320kbps, 43:11, 99.0 MiB
New: 1 items, MP3, 320kbps, 4:28, 10.2 MiB"

```
E:\Music\Paul & Linda McCartney\Ram (1 items)
Finding tags for album "Paul & Linda McCartney - Ram".
Candidates:
1. Paul McCartney & Linda McCartney - Ram (39.7%) (missing tracks, artist) (CD, 1988, GB, Capitol Records)
2. Paul McCartney & Linda McCartney - Ram (39.7%) (missing tracks, artist) (CD, 2012, GB, Hear Music, Paul McCartney Archive Collection remaster)
3. Paul McCartney & Linda McCartney - Ram (28.3%) (missing tracks, artist) (2xCD, 2012, US, Hear Music, Paul McCartney Archive Collection: Special Edition)
4. Paul McCartney & Linda McCartney - Ram (26.4%) (missing tracks, artist) (3xDigital Media, 2012, US, Hear Music, Paul McCartney Archive Collection: special edition)
5. Paul McCartney & Linda McCartney - Ram (7.9%) (missing tracks, artist) (11xCD, 2012, US, Hear Music, Paul McCartney Archive Collection: deluxe edition)
# selection (default 1), Skip, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort?
Correcting tags from:
    Paul & Linda McCartney - Ram
To:
    Paul McCartney & Linda McCartney - Ram
URL:
    https://musicbrainz.org/release/435b63d0-4e90-3695-a9ac-ca40b79c46f9
(Similarity: 39.7%) (missing tracks, artist) (CD, 1988, GB, Capitol Records)
 * The Back Seat Of My Car -> The Back Seat of My Car
Missing tracks (11/12 - 91.7%):
 ! Too Many People (#1) (4:10)
 ! 3 Legs (#2) (2:48)
 ! Ram On (#3) (2:30)
 ! Dear Boy (#4) (2:15)
 ! Uncle Albert / Admiral Halsey (#5) (4:54)
 ! Smile Away (#6) (3:53)
 ! Heart of the Country (#7) (2:24)
 ! Monkberry Moon Delight (#8) (5:25)
 ! Eat at Home (#9) (3:22)
 ! Long Haired Lady (#10) (6:04)
 ! Ram On (#11) (0:56)
Apply, More candidates, Skip, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort? a
This album is already in the library!
Old: 12 items, MP3, 320kbps, 43:11, 99.0 MiB
New: 1 items, MP3, 320kbps, 4:28, 10.2 MiB
[S]kip new, Keep both, Remove old? k
```

**Comment by sampsyo** [comment]
Hi! This sounds very close to #112, which proposes to let you merge existing tracks with new ones to get a more complete album. Does that sound like what you really want here?

**Comment by RollingStar** [comment]
The original issue for #112 is one I'm having. Other commenters here have cases where they have track 1-5 and then another folder they're trying to import with tracks 4-10. There's a conflict with what to do with tracks 4 and 5 that are tough to resolve automatically with a program.

However, I'm talking more about the automatching phase, before Beets even decides that there's a conflict. If my issue is fixed, #112 will still be an issue. It will just be easier to arrive at #112 without beets asking for input in the first automatch phase.

One way to do what I'm proposing (pseudocode):

```
if in.library(top.candidate) == TRUE {
    if in.library(candidate.tracks) == FALSE {
        candidate.score = candidate.score + 10%
    }
}
```

Naiively importing candidate tracks without considering the library in special cases results in beets being more conservative than it should be. It's not using all the information it has available -- namely, that the user already has some of a candidate album in their library, and it's missing tracks, *and* those missing tracks are the only candidates being imported.

**Comment by sampsyo** [comment]
Sure; that makes sense. Would you mind adding your thoughts to that thread for #112, though? I agree that there are easier and more difficult cases when completing an album—sometimes you have a perfect resolution, and sometimes you don't—but we should handle them all using the same flow. Specifically, we should use a general policy that asks whether merging the album would lead to a better fit with the candidate—that should cover your case ("easy" merging) as well as more complex ones.
